src/clientHandler.ts

```diff
@@ -0,0 +1,252 @@
+
+				if (!multiFoldersSupported) {
+					// restart is needed to launch folder-focused instances
+					console.log('Restarting clients as folder-focused');
+					await this.StopClients(folders);
+					this.supportsMultiFolders = false;
+					this.StartClients(folders);
+				}
+			});
+
+			disposables.push(tfClient.client.start());
+			clients.set("", tfClient);
+
+			return disposables;
+		}
+
+		if (folders && folders.length > 0) {
+			for (const folder of folders) {
+				if (!clients.has(folder)) {
+					console.log(`Starting client for ${folder}`);
+					const folderClient = this.newTerraformClient(folder);
+					folderClient.client.onReady().then(() => {
+						this.reporter.sendTelemetryEvent('startClient');
+					});
+
+					disposables.push(folderClient.client.start());
+					clients.set(folder, folderClient);
+				} else {
+					console.log(`Client for folder: ${folder} already started`);
+				}
+			}
+		}
+		return disposables;
+	}
+
+	private newTerraformClient(location?: string): TerraformLanguageClient {
+		const cmd = this.pathToBinary;
+		const binaryName = cmd.split('/').pop();
+
+		const serverArgs: string[] = config('terraform').get('languageServer.args');
+		const experimentalFeatures = config('terraform-ls').get('experimentalFeatures');
+
+		let channelName = `${binaryName}`;
+		let id = `languageServer`
+		let name = `Language Server`
+		let wsFolder: vscode.WorkspaceFolder;
+		let rootModulePaths: string[];
+		let excludeModulePaths: string[];
+		let documentSelector: DocumentSelector;
+		let outputChannel: vscode.OutputChannel;
+		if (location) {
+			channelName = `${binaryName}: ${location}`;
+			id = `languageServer/${location}`
+			name = `Language Server: ${location}`
+			wsFolder = getWorkspaceFolder(location);
+			documentSelector = [
+				{ scheme: 'file', language: 'terraform', pattern: `${wsFolder.uri.fsPath}/**/*` },
+				{ scheme: 'file', language: 'terraform-vars', pattern: `${wsFolder.uri.fsPath}/**/*` }
+			]
+			rootModulePaths = config('terraform-ls', wsFolder).get('rootModules');
+			excludeModulePaths = config('terraform-ls', wsFolder).get('excludeRootModules');
+			outputChannel = vscode.window.createOutputChannel(channelName);
+			outputChannel.appendLine(`Launching language server: ${cmd} ${serverArgs.join(' ')} for folder: ${location}`);
+		} else {
+			documentSelector = [
+				{ scheme: 'file', language: 'terraform' },
+				{ scheme: 'file', language: 'terraform-vars' }
+			]
+			rootModulePaths = config('terraform-ls').get('rootModules');
+			excludeModulePaths = config('terraform-ls').get('excludeRootModules');
+			outputChannel = vscode.window.createOutputChannel(channelName);
+			outputChannel.appendLine(`Launching language server: ${cmd} ${serverArgs.join(' ')}`);
+		}
+
+		if (rootModulePaths.length > 0 && excludeModulePaths.length > 0) {
+			throw new Error('Only one of rootModules and excludeRootModules can be set at the same time, please remove the conflicting config and reload');
```

Comment:
I personally wouldn't know how to reload 😄 Is there a restart language server command?

Reply:
So the server isn't even launched at this point - this would be reload of VS Code and interestingly we already have some logic to prompt the reload with a single button upon any change to the relevant parts of the configuration:
https://github.com/hashicorp/vscode-terraform/blob/583bdf59a19d46a0f5e2325f0d7a11379e0bd337/src/extension.ts#L108-L116

Now that you mention it though it looks like that the exception just blocks the process of creating a client and is never shown to the user - which is not ideal. That said this is not something being introduced as part of this PR, that bug was there for some time now (apparently since this error handling was introduced), so we can deal with that in a separate PR.